tools/brokerd/brokerd.cpp

```diff
@@ -0,0 +1,247 @@
+                    {"uptime",
+                     http::channel_container_entry{&broker::uptime_channel, &broker::uptime_cv}},
+                };
+
+                http::server (fs, &status->value (), channels,
+                              [announce_port] (in_port_t const port) {
+                                  if (announce_port) {
+                                      std::lock_guard<decltype (broker::iomut)> lock{broker::iomut};
+                                      std::cout << "HTTP listening on port " << port << std::endl;
+                                  }
+                              });
+            },
+            http_status.get ()));
+    }
+
+    // create worker threads
+    // ~~~~~~~~~~~~~~~~~~~~~
+    std::vector<std::future<void>> create_worker_threads (
+        std::shared_ptr<broker::command_processor> const & commands, brokerface::fifo_path & fifo,
+        std::shared_ptr<broker::scavenger> const & scav, std::atomic<bool> * const uptime_done) {
+
+        std::vector<std::future<void>> futures;
+
+        futures.push_back (create_thread ([&fifo, commands] () {
+            thread_init ("command");
+            commands->thread_entry (fifo);
+        }));
+
+        futures.push_back (create_thread ([scav] () {
+            thread_init ("scavenger");
+            scav->thread_entry ();
+        }));
+
+        futures.push_back (create_thread ([] () {
+            thread_init ("gcwatch");
+            broker::gc_process_watch_thread ();
+        }));
+
+        futures.push_back (create_thread (
+            [] (std::atomic<bool> * const done) {
+                thread_init ("uptime");
+                broker::uptime (done);
+            },
+            uptime_done));
+
+        return futures;
+    }
+
+    // make weak
+    // ~~~~~~~~~
+    /// Creates a weak_ptr from a shared_ptr. This can be done implicitly, but I want to make
+    /// the conversion explicit.
+    template <typename T>
+    std::weak_ptr<T> make_weak (std::shared_ptr<T> const & p) {
+        return {p};
+    }
+
+    // get http server status
+    // ~~~~~~~~~~~~~~~~~~~~~~
+    /// Create an HTTP server_status object which reflects the user's choice of port.
+    decltype (auto) get_http_server_status (maybe<in_port_t> const & port) {
+        if (port.has_value ()) {
+            return maybe<http::server_status>{in_place, *port};
+        }
+        return nothing<http::server_status> ();
+    }
+} // namespace
+
+namespace pstore {
+    namespace broker {
+        int run_broker (switches opt) {
+            using priority = logger::priority;
+
+            threads::set_name ("main");
+            create_log_stream ("broker.main");
+            log (priority::notice, "broker starting");
+
+#ifdef _WIN32
+            wsa_startup startup;
+            if (!startup.started ()) {
+                throw new std::exception ("WSAStartup() failed, broker exited");
```

Comment:
(FTR, I prefer the exception being raised here because it gives the caller a choice about how to handle the error.)

Reply:
Having said that, don’t throw a dynamically-allocated exception instance. Two reasons:

- The new could fail resulting in bad_alloc being raised.
- It would have to be caught by std::exception *: the caller is expecting to catch by value.

Also, it’s good practise to throw a subclass of std::exception so that a caller could distinguish by type.

Reply:
Made changes, thanks.